chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -325,4 +332,65 @@ Array PseudoInverse(const Array& a, float rcond) {
     return out;
 }
 
+std::tuple<Array, Array> Eigh(const Array& a, char uplo) {
+    CheckRankTwoArray(a);
+    CheckSquareMatrix(a);
+    CheckUplo(uplo);
+
+    Array w = Empty(Shape{a.shape()[0]}, a.dtype(), a.device());
+    Array v = Empty(a.shape(), a.dtype(), a.device());
+
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<SyevdKernel>(a, w, v, uplo, /*compute_v=*/true);
+        v = v.Transpose();
+    }
+
+    // Reference:
+    // Section 3.1 https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+    {
+        BackwardBuilder bb{"eigh", a, {w, v}};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([a_tok = bb.RetainInput(0), w_tok = bb.RetainOutput(0), v_tok = bb.RetainOutput(1)](BackwardContext& bctx) {
+                const Array& a = bctx.GetRetainedInput(a_tok);
```

Comment:
Let's not retain `a` just for the dtype and device. We can retain "a_type" only.